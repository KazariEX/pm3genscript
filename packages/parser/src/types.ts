import type { Block, Command, Dynamic, Identifier, Macro } from "./node";

export interface Diagnostic {
    message: string;
    offset: number;
    length: number;
}

export interface Token {
    type: TokenType;
    value: string;
    offset: number;
}

export interface TokenTypeMap {
    hash: "hash";
    at: "at";
    equal: "equal";
    identifier: "identifier";
    number: "number";
    string: "string";
}

export type TokenType = TokenTypeMap[keyof TokenTypeMap];

export interface AST {
    dynamic: {
        defines: Dynamic[];
        references: Dynamic[];
    };
    symbol: {
        defines: Macro[];
        references: Identifier[];
    };
    children: ASTChild[];
}

export type ASTChild = Block | Macro | Command;