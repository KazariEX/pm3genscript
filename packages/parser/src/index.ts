export * from "./check";
export * from "./node";
export * from "./parse";
export * from "./tokenize";
export * from "./types";
export * from "./utils/is";
export * from "./utils/traverse";