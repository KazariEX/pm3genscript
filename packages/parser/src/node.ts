export abstract class Node {
    constructor(
        public type: string,
        public offset: number
    ) {}

    getText(root: Root) {
        return root.text.slice(this.offset, this.getEnd());
    }

    abstract getEnd(): number;

    getLength() {
        return this.getEnd() - this.offset;
    }
}

export class Root extends Node {
    children: Node[] = [];

    constructor(
        public text: string
    ) {
        super("root", 0);
    }

    override getText() {
        return this.text;
    }

    getEnd() {
        return this.text.length;
    }
}

export abstract class Parent extends Node {
    arguments: Argument[] = [];

    constructor(
        public override type: string,
        public override offset: number,
        public name: Identifier
    ) {
        super(type, offset);
    }

    getEnd() {
        return this.arguments.length ? this.arguments.at(-1)!.getEnd() : this.name.getEnd();
    }
}

export class Macro extends Parent {
    constructor(
        public override offset: number,
        public override name: Identifier
    ) {
        super("macro", offset, name);
    }
}

export class Command extends Parent {
    constructor(
        public override name: Identifier
    ) {
        super("command", name.offset, name);
    }
}

export class Dynamic extends Node {
    constructor(
        public override offset: number,
        public name: Identifier
    ) {
        super("dynamic", offset);
    }

    getEnd() {
        return this.name.getEnd();
    }
}

export abstract class Literal extends Node {
    constructor(
        public override type: string,
        public override offset: number,
        public value: string
    ) {
        super(type, offset);
    }

    override getText() {
        return this.value;
    }

    getEnd() {
        return this.offset + this.value.length;
    }

    override getLength() {
        return this.value.length;
    }
}

export class Identifier extends Literal {
    constructor(
        public override offset: number,
        public override value: string
    ) {
        super("identifier", offset, value);
    }
}

export class NumberLiteral extends Literal {
    constructor(
        public override offset: number,
        public override value: string
    ) {
        super("number", offset, value);
    }
}

export class StringLiternal extends Literal {
    constructor(
        public override offset: number,
        public override value: string
    ) {
        super("string", offset, value);
    }
}

export type Argument = Dynamic | Identifier | NumberLiteral | StringLiternal;

export class Block extends Node {
    children: Command[] = [];

    constructor(
        public label: Macro
    ) {
        super("block", label.offset);
    }

    getEnd() {
        return this.children.length ? this.children.at(-1)!.getEnd() : this.label.getEnd();
    }
}