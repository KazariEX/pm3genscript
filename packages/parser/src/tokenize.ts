import type { Diagnostic, Token, TokenType } from "./types";

export function tokenize(text: string) {
    const tokens: Token[] = [];
    const diagnostics: Diagnostic[] = [];
    const textLength = text.length;
    let idx = 0;

    while (idx < textLength) {
        const char = text[idx];
        switch (char) {
            case "#": {
                add("hash", char, idx);
                idx++;
                break;
            }
            case "@": {
                add("at", char, idx);
                idx++;
                break;
            }
            case "=": {
                add("equal", char, idx);
                idx++;
                break;
            }
            default: {
                const offset = idx;
                let value = "";
                let next = char;
                if (isWhitespace(char)) {
                    do {
                        next = text[++idx];
                    } while (idx < textLength && isWhitespace(next));
                    break;
                }
                else if (isUppercaseAlpha(char)) {
                    do {
                        value += next;
                        next = text[++idx];
                    } while (idx < textLength && isUppercaseAlpha(next) || isNumeric(next) || next === "_");
                    add("symbol", value, offset);
                    break;
                }
                else if (isLowercaseAlpha(char)) {
                    do {
                        value += next;
                        next = text[++idx];
                    } while (idx < textLength && (isLowercaseAlpha(next) || isNumeric(next) || next === "_"));
                    add("identifier", value, offset);
                    break;
                }
                else if (isNumeric(char)) {
                    let isHex = false;
                    do {
                        value += next;
                        next = text[++idx];
                        if (value.length === 1 && next === "x") {
                            isHex = true;
                            value += next;
                            next = text[++idx];
                        }
                    } while (idx < textLength && (isHex ? isHexNumeric(next) : isNumeric(next)));
                    add("number", value, offset);
                    break;
                }
                else if (char === "\"" || char === "'") {
                    const quote = char;
                    do {
                        value += next;
                        next = text[++idx];
                    } while (idx < textLength && next !== quote);
                    value += next;
                    idx++;
                    add("string", value, offset);
                    break;
                }
                diagnostics.push({
                    message: `Unexpected character "${char}".`,
                    offset: idx,
                    length: 1
                });
                idx++;
            }
        }
    }

    return {
        tokens,
        diagnostics
    };

    function add(type: TokenType, value: string, offset: number) {
        tokens.push({
            type,
            value,
            offset
        });
    }
}

function isWhitespace(char: string) {
    return char === " " || char === "\n" || char === "\r" || char === "\t";
}

function isUppercaseAlpha(char: string) {
    const code = char.charCodeAt(0);
    return code >= 65 && code <= 90;
}

function isLowercaseAlpha(char: string) {
    const code = char.charCodeAt(0);
    return code >= 97 && code <= 122;
}

function isNumeric(char: string) {
    const code = char.charCodeAt(0);
    return code >= 48 && code <= 57;
}

function isHexNumeric(char: string) {
    const code = char.charCodeAt(0);
    return code >= 97 && code <= 102
        || code >= 65 && code <= 70
        || code >= 48 && code <= 57;
}