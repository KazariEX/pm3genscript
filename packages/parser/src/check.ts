import type { ArgumentType } from "@pm3genscript/shared";
import { type Argument, Block, type Command, Dynamic, Identifier, Macro, NumberLiteral, type Parent, StringLiternal } from "./node";
import type { AST, ASTChild, Diagnostic } from "./types";

export const symbolRE = /^[A-Z][A-Z0-9_]*$/;

export function check(ast: AST) {
    const diagnostics: Diagnostic[] = [];
    const resolvedValues = new Map<Argument, any>();
    const dynamicNames = new Map<string, {
        identifier: Identifier;
        references: Identifier[];
    }>();
    const symbols = new Map<string, {
        identifier: Exclude<Argument, Dynamic>;
        references: Identifier[];
        type: ArgumentType;
        value: any;
    }>();

    for (const macro of ast.symbol.defines) {
        if (macro.arguments.length < 2) {
            continue;
        }
        const [arg0, arg1] = macro.arguments;
        if (arg0 instanceof Identifier && symbolRE.test(arg0.value)) {
            const name = arg0.value;
            const [type, value] = resolveRuntimeTypeAndValue(arg1, false);
            if (symbols.has(name)) {
                symbols.get(name)!.identifier = arg0;
            }
            else {
                symbols.set(name, {
                    identifier: arg0,
                    references: [],
                    type,
                    value
                });
            }
        }
    }
    for (const reference of ast.symbol.references) {
        const name = reference.value;
        if (symbols.has(name)) {
            symbols.get(name)!.references.push(reference);
        }
        else {
            diagnostics.push({
                message: `Symbol "${name}" is not defined.`,
                offset: reference.offset,
                length: reference.getLength()
            });
        }
    }

    for (const dynamic of ast.dynamic.defines) {
        const name = dynamic.name.value;
        if (!dynamicNames.has(name)) {
            dynamicNames.set(name, {
                identifier: dynamic.name,
                references: []
            });
        }
        else {
            diagnostics.push({
                message: `Dynamic offset "@${name}" is already defined.`,
                offset: dynamic.offset,
                length: dynamic.getLength()
            });
        }
    }
    for (const reference of ast.dynamic.references) {
        const name = reference.name.value;
        if (dynamicNames.has(name)) {
            dynamicNames.get(name)!.references.push(reference.name);
        }
        else {
            diagnostics.push({
                message: `Dynamic offset "@${name}" is not defined.`,
                offset: reference.offset,
                length: reference.getLength()
            });
        }
    }

    for (const child of forEachChild(ast.children)) {
        if (child instanceof Macro) {
            checkMacro(child);
        }
        else {
            checkCommand(child);
        }
    }

    return {
        diagnostics,
        resolvedValues,
        dynamicNames,
        symbols
    };

    function checkMacro(macro: Macro) {
        const { template } = macro;
        if (!template) {
            diagnostics.push({
                message: `Unknown macro "${macro.name.value}".`,
                offset: macro.name.offset,
                length: macro.name.getLength()
            });
            return;
        }

        const { arguments: args } = macro;
        const { arguments: templateArgs = [] } = template;
        const length = Math.max(args.length, templateArgs.length);

        for (let i = 0; i < length; i++) {
            const arg = args[i];
            const templateArg = templateArgs[i];
            if (!arg || !templateArg) {
                reportArgumentCountMismatch(arg, args, templateArgs.length, macro);
                break;
            }

            const allowedTypes = getAllowedTypes(templateArg.type ?? []);
            const canSymbol = (templateArg.can?.symbol ?? true) && !allowedTypes.has("symbol");
            const [type, value] = resolveRuntimeTypeAndValue(arg, canSymbol);

            if (templateArg.enum) {
                if (!templateArg.enum.includes(value)) {
                    diagnostics.push({
                        message: `Value "${value}" is not allowed for argument "${templateArg.name}".`,
                        offset: arg.offset,
                        length: arg.getLength()
                    });
                }
            }
            else if (!allowedTypes.has(type)) {
                diagnostics.push({
                    message: `Expected argument type "${templateArg.type}", got "${type}".`,
                    offset: arg.offset,
                    length: arg.getLength()
                });
            }
        }
    }

    function checkCommand(command: Command) {
        const { arguments: args, template } = command;
        const { arguments: templateArgs = [] } = template;
        const length = Math.max(args.length, templateArgs.length);
        const values: any[] = [];

        for (let i = 0; i < length; i++) {
            const arg = args[i];
            const templateArg = templateArgs[i];
            if (!arg || !templateArg) {
                reportArgumentCountMismatch(arg, args, templateArgs.length, command);
                break;
            }

            const allowedTypes = getAllowedTypes(templateArg.type ?? []);
            const [type, value] = resolveRuntimeTypeAndValue(arg, !allowedTypes.has("symbol"));

            if (templateArg.when && !templateArg.when?.(values)) {
                break;
            }
            values.push(value);

            if (!allowedTypes.has(type)) {
                diagnostics.push({
                    message: `Expected argument type "${templateArg.type}", got "${type}".`,
                    offset: arg.offset,
                    length: arg.getLength()
                });
            }
        }
    }

    function reportArgumentCountMismatch(arg: Argument, args: Argument[], templateArgLength: number, parent: Parent) {
        const [start, end] = arg
            ? [arg.offset, args.at(-1)!.getEnd()]
            : [parent.offset, parent.getEnd()];
        diagnostics.push({
            message: `Expected ${templateArgLength} argument(s), got ${args.length}.`,
            offset: start,
            length: end - start
        });
    }

    function resolveRuntimeTypeAndValue(arg: Argument, canSymbol = true): [ArgumentType, any] {
        if (arg instanceof Dynamic) {
            const value = arg.name.value;
            resolvedValues.set(arg, value);
            return ["pointer", value];
        }
        if (arg instanceof Identifier) {
            const value = arg.value;
            if (symbolRE.test(value)) {
                if (canSymbol) {
                    const symbol = symbols.get(value);
                    if (symbol) {
                        return [symbol.type, symbol.value];
                    }
                }
                return ["symbol", value];
            }
            resolvedValues.set(arg, value);
            return ["identifier", value];
        }
        if (arg instanceof NumberLiteral) {
            const value = arg.value.startsWith("0x")
                ? Number.parseInt(arg.value.slice(2), 16)
                : Number.parseInt(arg.value);
            resolvedValues.set(arg, value);

            const type = value >= 0 && value <= 0xFF ? "byte"
                : value >= 0 && value <= 0xFFFF ? "word"
                : value >= 0 && value <= 0xFFFFFFFF ? "dword"
                : "number";
            return [type, value];
        }
        if (arg instanceof StringLiternal) {
            const value = arg.value;
            resolvedValues.set(arg, value);
            return ["string", value];
        }
        else {
            throw arg satisfies never;
        }
    }
}

function* forEachChild(children: ASTChild[]): Generator<Macro | Command> {
    for (const child of children) {
        if (child instanceof Block) {
            yield child.label;
            yield* child.children;
        }
        else {
            yield child;
        }
    }
}

function getAllowedTypes(type: ArgumentType | ArgumentType[]) {
    const types = Array.isArray(type) ? type : [type];
    if (types.includes("number") || types.includes("dword") || types.includes("pointer")) {
        types.push("byte", "word", "dword", "pointer");
    }
    else if (types.includes("word")) {
        types.push("byte");
    }
    return new Set(types);
}