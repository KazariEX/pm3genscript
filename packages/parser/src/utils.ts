import { Block, Command, Dynamic, Macro, type Node, Parent, Root } from "./node";

export function forEachNode(
    node: Node,
    visit: (node: Node, parent: Node | undefined, parents: Node[]) => void
) {
    const parents: Node[] = [];
    traverse(node);

    function traverse(node: Node) {
        visit(node, parents[parents.length - 1], parents);

        parents.push(node);
        if (node instanceof Root) {
            for (const child of node.children) {
                traverse(child);
            }
        }
        else if (node instanceof Block) {
            traverse(node.label);
            for (const child of node.children) {
                traverse(child);
            }
        }
        else if (node instanceof Parent) {
            traverse(node.name);
            for (const arg of node.arguments) {
                traverse(arg);
            }
        }
        else if (node instanceof Dynamic) {
            traverse(node.name);
        }
        parents.pop();
    }
}

export function* forEachStatement(children: Node[]): Generator<Macro | Command> {
    for (const child of children) {
        if (child instanceof Block) {
            yield child.label;
            yield* child.children;
        }
        else if (child instanceof Macro || child instanceof Command) {
            yield child;
        }
    }
}