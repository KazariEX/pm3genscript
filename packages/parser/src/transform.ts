import { symbolRE } from "./check";
import { Block, Command, Dynamic, Identifier, Macro, type Root } from "./node";
import type { AST, Diagnostic } from "./types";

export function transform(root: Root) {
    const ast: AST = {
        dynamic: {
            defines: [],
            references: []
        },
        symbol: {
            defines: [],
            references: []
        },
        children: []
    };
    const diagnostics: Diagnostic[] = [];
    let currentBlock: Block | null = null;

    for (const child of root.children) {
        if (child instanceof Macro) {
            processMacro(child);
        }
        else if (child instanceof Command) {
            processCommand(child);
        }
        else {
            diagnostics.push({
                message: `Expected "macro" or "command" node at the root, got "${child.type}"`,
                offset: child.offset,
                length: child.getLength()
            });
        }
    }

    return {
        ast,
        diagnostics
    };

    function processMacro(macro: Macro) {
        switch (macro.canonicalName) {
            case "org": {
                const block = new Block(macro);
                currentBlock = block;
                ast.children.push(block);

                const arg0 = macro.arguments[0];
                if (arg0 instanceof Dynamic) {
                    ast.dynamic.defines.push(arg0);
                }
                break;
            }
            case "define": {
                ast.symbol.defines.push(macro);
            }
            default: {
                currentBlock = null;
                ast.children.push(macro);
                break;
            }
        }
    }

    function processCommand(command: Command) {
        for (const arg of command.arguments) {
            if (arg instanceof Dynamic) {
                ast.dynamic.references.push(arg);
            }
            else if (arg instanceof Identifier && symbolRE.test(arg.value)) {
                ast.symbol.references.push(arg);
            }
        }

        if (currentBlock) {
            currentBlock.children.push(command);
        }
        else {
            ast.children.push(command);
            diagnostics.push({
                message: `Command "${command.name.value}" is not inside a block.`,
                offset: command.offset,
                length: command.getLength()
            });
        }
    }
}