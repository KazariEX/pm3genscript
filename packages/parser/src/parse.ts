import { commands } from "@pm3genscript/shared";
import { Block, Command, Dynamic, Identifier, Macro, type Node, NumberLiteral, Root, StringLiternal, Symbol } from "./node";
import { tokenize } from "./tokenize";
import { isCommand } from "./utils/is";
import type { Diagnostic, Token } from "./types";

export function parse(text: string) {
    const tokenized = tokenize(text);
    const walked = walkTokens(text, tokenized.tokens);

    return {
        ast: walked.root,
        diagnostics: [
            ...tokenized.diagnostics,
            ...walked.diagnostics
        ]
    };
}

export function walkTokens(text: string, tokens: Token[]) {
    const root = new Root(text);
    const diagnostics: Diagnostic[] = [];
    const tokenLength = tokens.length;
    let current = 0;

    const parents: (Macro | Command)[] = [];
    let currentBlock: Block | null = null;

    root: while (current < tokenLength) {
        const token = tokens[current];

        switch (token.type) {
            case "hash": {
                let name: Identifier;
                const next = advance();
                if (next?.type === "identifier") {
                    name = new Identifier(next.offset, next.value);
                    current++;
                }
                else {
                    name = new Identifier(token.offset + token.value.length, "");
                    diagnostics.push({
                        message: `Expected identifier after "#", got "${next.type}".`,
                        offset: next.offset,
                        length: next.value.length
                    });
                }
                const macro = new Macro(token.offset, name);
                if (macro.canonicalName === "org") {
                    const block = new Block(macro);
                    currentBlock = block;
                    root.children.push(block);
                }
                else {
                    currentBlock = null;
                    root.children.push(macro);
                }
                parents.unshift(macro);
                if (macro.canonicalName === "break") {
                    break root;
                }
                break;
            }
            case "at": {
                let name: Identifier;
                const next = advance();
                if (next?.type === "identifier" || next?.type === "number") {
                    name = new Identifier(next.offset, next.value);
                    current++;
                }
                else {
                    name = new Identifier(token.offset + token.value.length, "");
                    diagnostics.push({
                        message: `Expected identifier or number after "@", got "${next.type}".`,
                        offset: next.offset,
                        length: next.value.length
                    });
                }
                const dynamic = new Dynamic(token.offset, name);
                attach(dynamic);
                break;
            }
            case "equal":
            case "identifier": {
                const identifier = new Identifier(token.offset, token.value);
                if (token.type === "equal" || identifier.value in commands) {
                    const command = new Command(identifier);
                    parents.unshift(command);
                    attach(command);
                }
                else {
                    attach(identifier);
                }
                current++;
                break;
            }
            case "symbol": {
                const symbol = new Symbol(token.offset, token.value);
                attach(symbol);
                current++;
                break;
            }
            case "number": {
                const number = new NumberLiteral(token.offset, token.value);
                attach(number);
                current++;
                break;
            }
            case "string": {
                const string = new StringLiternal(token.offset, token.value);
                attach(string);
                current++;
                break;
            }
        }
    }

    return {
        root,
        diagnostics
    };

    function advance() {
        return tokens[++current];
    }

    function attach(node: Node) {
        const container = isCommand(node) ? currentBlock?.children : parents[0]?.arguments;
        (container ?? root.children).push(node);
    }
}