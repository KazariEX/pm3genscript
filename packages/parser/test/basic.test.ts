import { describe, expect, it } from "vitest";
import { walkTokens } from "../src/parse";
import { tokenize } from "../src/tokenize";

describe("basic", () => {
    const text = `
#dynamic 0xA00000

#org @1
msgbox @2 MSG_FACE
end

#org @2
= "你自己魔怔吧，我要去二次元了。"
`.trim();

    const tokenized = tokenize(text);
    const walked = walkTokens(text, tokenized.tokens);

    it("tokenize", () => {
        expect(tokenized.diagnostics.length).toEqual(0);
        expect(tokenized.tokens).toMatchSnapshot();
    });

    it("walk", () => {
        expect(walked.diagnostics.length).toEqual(0);
        expect(walked.root).toMatchSnapshot();
    });
});