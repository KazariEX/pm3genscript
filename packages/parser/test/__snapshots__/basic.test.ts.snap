// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic > tokenize 1`] = `
[
  {
    "offset": 0,
    "type": "hash",
    "value": "#",
  },
  {
    "offset": 1,
    "type": "identifier",
    "value": "dynamic",
  },
  {
    "offset": 9,
    "type": "number",
    "value": "0xA00000",
  },
  {
    "offset": 19,
    "type": "hash",
    "value": "#",
  },
  {
    "offset": 20,
    "type": "identifier",
    "value": "org",
  },
  {
    "offset": 24,
    "type": "at",
    "value": "@",
  },
  {
    "offset": 25,
    "type": "number",
    "value": "1",
  },
  {
    "offset": 27,
    "type": "identifier",
    "value": "msgbox",
  },
  {
    "offset": 34,
    "type": "at",
    "value": "@",
  },
  {
    "offset": 35,
    "type": "number",
    "value": "2",
  },
  {
    "offset": 37,
    "type": "identifier",
    "value": "MSG_FACE",
  },
  {
    "offset": 46,
    "type": "identifier",
    "value": "end",
  },
  {
    "offset": 51,
    "type": "hash",
    "value": "#",
  },
  {
    "offset": 52,
    "type": "identifier",
    "value": "org",
  },
  {
    "offset": 56,
    "type": "at",
    "value": "@",
  },
  {
    "offset": 57,
    "type": "number",
    "value": "2",
  },
  {
    "offset": 59,
    "type": "equal",
    "value": "=",
  },
  {
    "offset": 61,
    "type": "string",
    "value": ""你自己魔怔吧，我要去二次元了。"",
  },
]
`;

exports[`basic > walk 1`] = `
Root {
  "children": [
    Macro {
      "arguments": [
        NumberLiteral {
          "offset": 9,
          "type": "number",
          "value": "0xA00000",
        },
      ],
      "canonicalName": "dynamic",
      "name": Identifier {
        "offset": 1,
        "type": "identifier",
        "value": "dynamic",
      },
      "offset": 0,
      "type": "macro",
    },
    Macro {
      "arguments": [
        Dynamic {
          "name": Identifier {
            "offset": 25,
            "type": "identifier",
            "value": "1",
          },
          "offset": 24,
          "type": "dynamic",
        },
      ],
      "canonicalName": "org",
      "name": Identifier {
        "offset": 20,
        "type": "identifier",
        "value": "org",
      },
      "offset": 19,
      "type": "macro",
    },
    Command {
      "arguments": [
        Dynamic {
          "name": Identifier {
            "offset": 35,
            "type": "identifier",
            "value": "2",
          },
          "offset": 34,
          "type": "dynamic",
        },
        Identifier {
          "offset": 37,
          "type": "identifier",
          "value": "MSG_FACE",
        },
      ],
      "canonicalName": "msgbox",
      "name": Identifier {
        "offset": 27,
        "type": "identifier",
        "value": "msgbox",
      },
      "offset": 27,
      "type": "command",
    },
    Command {
      "arguments": [],
      "canonicalName": "end",
      "name": Identifier {
        "offset": 46,
        "type": "identifier",
        "value": "end",
      },
      "offset": 46,
      "type": "command",
    },
    Macro {
      "arguments": [
        Dynamic {
          "name": Identifier {
            "offset": 57,
            "type": "identifier",
            "value": "2",
          },
          "offset": 56,
          "type": "dynamic",
        },
      ],
      "canonicalName": "org",
      "name": Identifier {
        "offset": 52,
        "type": "identifier",
        "value": "org",
      },
      "offset": 51,
      "type": "macro",
    },
    Command {
      "arguments": [
        StringLiternal {
          "offset": 61,
          "type": "string",
          "value": ""你自己魔怔吧，我要去二次元了。"",
        },
      ],
      "canonicalName": "=",
      "name": Identifier {
        "offset": 59,
        "type": "identifier",
        "value": "=",
      },
      "offset": 59,
      "type": "command",
    },
  ],
  "offset": 0,
  "text": "#dynamic 0xA00000

#org @1
msgbox @2 MSG_FACE
end

#org @2
= "你自己魔怔吧，我要去二次元了。"",
  "type": "root",
}
`;
